ibc-core/ics02-client/src/handler/recover_client.rs

```diff
@@ -0,0 +1,83 @@
+//! Protocol logic for processing ICS02 messages of type `MsgRecoverClient`.
+
+use ibc_core_client_context::prelude::*;
+use ibc_core_client_types::error::ClientError;
+use ibc_core_client_types::msgs::MsgRecoverClient;
+use ibc_core_handler_types::error::ContextError;
+use ibc_core_host::{ExecutionContext, ValidationContext};
+
+/// Performs the validation steps associated with the client recovery process. This
+/// includes validating that the parameters of the subject and substitute clients match,
+/// as well as validating that the substitute client *is* active and that the subject
+/// client is *not* active.
+pub fn validate<Ctx>(ctx: &Ctx, msg: MsgRecoverClient) -> Result<(), ContextError>
+where
+    Ctx: ValidationContext,
+{
+    let signer = msg.signer;
+    let subject_client_id = msg.subject_client_id.clone();
+    let substitute_client_id = msg.substitute_client_id.clone();
+
+    ctx.validate_message_signer(&signer)?;
+
+    let client_val_ctx = ctx.get_client_validation_context();
+
+    let subject_client_state = client_val_ctx.client_state(&subject_client_id)?;
+    let substitute_client_state = client_val_ctx.client_state(&substitute_client_id)?;
+
+    let subject_height = subject_client_state.latest_height();
+    let substitute_height = substitute_client_state.latest_height();
+
+    if subject_height >= substitute_height {
+        return Err(ClientError::ClientRecoveryHeightMismatch {
+            subject_height,
+            substitute_height,
+        }
+        .into());
+    }
```

Comment:
On a side note: should we start using something like [ensure](https://docs.rs/eyre/latest/eyre/macro.ensure.html) (or a util function for our usecases)? I feel that `ensure` trick is more readable.